### PR TITLE
Add support for `wolfSSL_set_verify` methods

### PR DESCRIPTION
**Pull request by kp-mariappan-ramasamy**
Add api's for calling `wolfSSL_set_verify` and `wolfSSL_CTX_set_verify` methods. For simplicity, set the last `VerifyMethod` callback parameter in both the apis as None. We can add it in future if requires.

### CODE REVIEW

File: wolfssl/src/chacha20_poly1305.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+/// and stores the generated authentication tag in the output buffer
+pub fn chacha20_poly1305_encrypt(
+    plain_text: &[u8],
+    key: ChaCha20Poly1305AeadKey,
+    iv: ChaCha20Poly1305AeadIv,
+) -> Result<(BytesMut, ChaCha20Poly1305AeadAuthTag), String> {
+    let mut cipher_text = BytesMut::with_capacity(plain_text.len());
+    let mut auth_tag = [0u8; 16];
+
+    // SAFETY: Will add later
```
Comment (xv-ian-c) [inline wolfssl/src/chacha20_poly1305.rs]: Don't forget ;-) (and below).

**Review by xv-ian-c — COMMENTED** [review]
Seems fine apart from the missing safety comments (but it's still draft so I guess that's ok for now)

File: wolfssl/src/chacha20_poly1305.rs (hunk 1 above)
Comment (kp-mariappan-ramasamy) [reply wolfssl/src/chacha20_poly1305.rs]: Yes ofcourse

File: wolfssl/src/chacha20_poly1305.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        // - `inKey` should be a valid pointer with size `wolfssl::CHACHA20_POLY1305_AEAD_KEYSIZE`
+        // - `inIv` should be a valid pointer with size `wolfssl::CHACHA20_POLY1305_AEAD_IV_SIZE`
+        // - `outAuthTag` should be a valid mutable pointer with size equal to `wolfssl::CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE`
+        // - `inAAD` should be a valid pointer with size equal to `inAADLen`
+        // - `inPlaintext` should be a valid pointer with size equal to `inPlaintextLen`
+        // - `outCipherText` should be a valid mutable pointer with size equal to `inPlaintextLen`
+        //
+        // First three invariants are maintained using byteslice of respective sizes.
+        // We do not use `inAAD`, hence it is null and 0.
+        // `inPlaintext` is again a bytes slice and `inPlaintextLen` is using len() of bytesslice
```
Comment (xv-ian-c) [inline wolfssl/src/chacha20_poly1305.rs:46]: nit "byteslice"

File: wolfssl/src/chacha20_poly1305.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    /// This function encrypts an input message `plain_text`, using the ChaCha20 stream cipher,
+    /// It also performs Poly-1305 authentication (on the cipher text),
+    /// and stores the generated authentication tag in the output buffer
+    pub fn encrypt(
+        &self,
+        iv: [u8; Self::IV_SIZE],
+        plain_text: &[u8],
+    ) -> Result<(BytesMut, [u8; Self::AUTHTAG_SIZE]), String> {
+        let mut cipher_text = BytesMut::zeroed(plain_text.len());
```
Comment (xv-ian-c) [inline wolfssl/src/chacha20_poly1305.rs:33]: You could use `BytesMut::with_capacity()` here and then [`cipher_text.spare_capacity_mut`](https://docs.rs/bytes/latest/bytes/struct.BytesMut.html#method.spare_capacity_mut) here and avoid bothering to initialize the input buffer just to be overwritten in the FFI call.

`Session<IOCB>::try_read` has an example.

File: wolfssl/src/chacha20_poly1305.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    /// an authentication generated with the inAAD (arbitrary length additional authentication data).
+    /// Note: If the generated authentication tag does not match the supplied authentication tag,
+    /// the text is not decrypted.
+    pub fn decrypt(
+        &self,
+        iv: [u8; Self::IV_SIZE],
+        cipher_text: &[u8],
+        auth_tag: [u8; Self::AUTHTAG_SIZE],
+    ) -> Result<BytesMut, String> {
+        let mut plain_text = BytesMut::zeroed(cipher_text.len());
```
Comment (xv-ian-c) [inline wolfssl/src/chacha20_poly1305.rs:77]: as above

File: wolfssl/src/chacha20_poly1305.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        // - `inKey` should be a valid pointer with size `wolfssl::CHACHA20_POLY1305_AEAD_KEYSIZE`
+        // - `inIv` should be a valid pointer with size `wolfssl::CHACHA20_POLY1305_AEAD_IV_SIZE`
+        // - `inAuthTag` should be a valid pointer with size `wolfssl::CHACHA20_POLY1305_AEAD_AUTHTAG_SIZE`
+        // - `inAAD` should be a valid pointer with size equal to `inAADLen`
+        // - `inCiphertext` should be a valid pointer with size equal to `inCiphertextLen`
+        // - `outPlainText` should be a valid mutable pointer with size equal to `inCiphertextLen`
+        //
+        // First three invariants are maintained using byteslice of respective sizes.
+        // We do not use `inAAD`, hence it is null and 0.
+        // `inCiphertext` is again a bytes slice and `inCiphertextLen` is using len() of bytesslice
```
Comment (xv-ian-c) [inline wolfssl/src/chacha20_poly1305.rs:89]: same nit

**Review by xv-ian-c — DISMISSED** [review]

File: wolfssl/src/chacha20_poly1305.rs (hunk 3 above)
Comment (kp-mariappan-ramasamy) [reply wolfssl/src/chacha20_poly1305.rs]: `BytesMut::spare_capacity_mut` needed unsafe code to set_len after populating with data, so avoided it. But who am i kidding.
It's already unsafe to call C apis and it is in data path. Updated it to use `spare_capacity_mut`

**Review by xv-ian-c — APPROVED** [review]
LGTM. Would be great to have some tests for the chacha20 stuff.

**Comment by kp-mariappan-ramasamy** [comment]
Added tests in this PR:
https://github.com/expressvpn/wolfssl-rs/pull/120

**Pull request merged**